api/envoy/service/discovery/v3/discovery.proto

```diff
@@ -230,6 +275,14 @@ message DeltaDiscoveryResponse {
   // [#not-implemented-hide:]
   // The control plane instance that sent the response.
   config.core.v3.ControlPlane control_plane = 7;
+
+  // [#not-implemented-hide:]
+  // An optional per-resource status response that can be used when no matching resource(s)
+  // are sent to the client, and will describe the reason. This will be a map
+  // from a resource name to its status. A resource that doesn't appear in the map
+  // implies the resource response is valid. No status map implies that all the
+  // resources in the response are valid.
+  map<string, DiscoveryResponseStatus> resources_status = 8;
```

Comment:
Note that unlike the SotW protocol, which currently doesn't have a way to explicitly indicate that a subscribed resource does not exist, the delta protocol *does* have such a mechanism (the `removed_resources` field).  So in this case, the only value of this field is to support the version-mismatch case.  And if we handle the version via flexible range-based context parameter matching, then we wouldn't need this field at all.

Reply:
> And if we handle the version via flexible range-based context parameter matching, then we wouldn't need this field at all.

How does this relate? The management server can still be missing a version in the specified range.

Reply:
Right, but if the version was just a context parameter, then it would effectively be part of the name, so `NO_MATCHING_RESOURCE` and `NO_MATCHING_VERSION_RESOURCE` would be the same thing.  And at that point, we would not need this field, because the existing `removed_resources` field would do what we need.

Reply:
In this case, we've actually strictly lost information, but I agree it's a more consistent and logical way of viewing things. I think @mattklein123 was interested in a specific "no matching version" feedback signal, but maybe this would be sufficient.

Reply:
I think the requirement here is that we have a clear way of providing status to the client about why a resource was not returned. The `DiscoveryResponseStatus` message seems to do that nicely. I don't have a strong preference on how we do it exactly but I agree with @htuch that we should not lose information on why something happened and should probably have it look similar for both SoTW and delta, whatever the best way of accomplishing that is.

Reply:
To be clear, I certainly don't object to having a way of exposing a more detailed status to the client; I think that's definitely a good thing for debuggability.  I'm more concerned about having two different ways of doing the same thing.  An alternative here would be to simply deprecate the existing `removed_resources` field and replace it with this.

Also, could we consider using `google.rpc.Status` instead of a custom `DiscoveryResponseStatus` enum?  That seems a bit more flexible in the long run.

Reply:
I'm fine with whatever solution you all agree to on this as long as we maintain the data fidelity.

Reply:
> Right, but if the version was just a context parameter, then it would effectively be part of the name, so `NO_MATCHING_RESOURCE` and `NO_MATCHING_VERSION_RESOURCE` would be the same thing. And at that point, we would not need this field, because the existing `removed_resources` field would do what we need.

Maybe I got it wrong, but I thought of the response status as a complimentary field to the removed resources, providing the reason why were they removed. It essentially allows the following differentiation: a requested resource name (no matter what version) doesn't exist in the management server; and a requested resource in a specific version range doesn't exist in the management server (the server only handles older vs newer versions).
At the moment, the client doesn't use this information (other than logging) but it does allow the flexibility to request the resource from other servers, or to notify the operator that there's an issue that needs to be addressed (the client needs to be updated).

 > Also, could we consider using `google.rpc.Status` instead of a custom `DiscoveryResponseStatus` enum? That seems a bit more flexible in the long run.

I agree it should be used, and stating that the _code_ field in it will be one of the `DiscoveryResponseStatusTypes` that we define.

Reply:
I don't feel strongly about it, but I tend to agree with Mark that resources_status could be a superset of removed_resources and we can just build it in a way that would allow differentiating the two cases?

Reply:
> At the moment, the client doesn't use this information (other than logging) but it does allow the flexibility to request the resource from other servers, or to notify the operator that there's an issue that needs to be addressed (the client needs to be updated).

I suspect that there are no cases where the client will need to programmatically differentiate between these two cases.  Knowing the difference is useful for humans but not really helpful for client implementations.  Most clients will simply treat the resource as non-existing in both cases; I could imagine a client deciding to try another management server instead, but it seems unlikely that a client would choose to do so in one of these two cases but not the other.

That's why I favor simply deprecating the existing `removed_resources` field and replacing it with `map<string, google.rpc.Status> resources_not_available`, where the map value would be an error message indicating why the specified resource was not available.  This approach would allow us to handle any arbitrary reason why the resource is not available, in case we run into cases other than the two we are currently considering.

> I agree it should be used, and stating that the _code_ field in it will be one of the `DiscoveryResponseStatusTypes` that we define.

I think the code in that proto needs to be one of the standard status code enum values.  It would be very surprising to users for the code to be a new error space.

Reply:
Yeah, the primary requirement is debuggability; as long as we can include the version failure in the `google.rpc.Status` string it's all good.